### PR TITLE
Feature/caching mechanism

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/OAuth2ApiClient.js [hunk 1]
```diff
@@ -9,10 +9,9 @@ class OAuth2ApiClient extends ApiClient{
     async callApi(path, httpMethod, pathParams, queryParams, headerParams, formParams, bodyParam, authNames, contentTypes, accepts, returnType) {
 
         let getTokenResponseResult = await this.authService.getTokenResponse();
```
Comment (sfdrogojan) [inline src/OAuth2ApiClient.js]: Can getTokenResponseResult be renamed tokenResponseResult?

File: test/auth/AuthServiceUnitTests.spec.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
+const CacheService = require('../../src/Auth/CacheService');
+const TestHelper = require('./TestHelper');
+const ApiClient = require('../../src/ApiClient');
+const AuthService = require('../../src/Auth/AuthService');
+
+const expect = require('expect.js');
+const sinon = require('sinon');
+
+const date = 1568972613184;
```
Comment (sfdrogojan) [inline test/auth/AuthServiceUnitTests.spec.js]: For more readability, you can use for example new Date(2000, 1, 1).getTime()

File: test/auth/AuthServiceUnitTests.spec.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    describe('getTokenResponse', function () {
+        it('should execute the callApi method when there is no cached value for a cache key', async () => {
+            let authService = new AuthService(clientConfiguration, apiClientStub, cacheService);
+
+            await authService.getTokenResponse();
+
+            expect(apiClientStub.callApi.calledOnce).to.eql(true);
+            }
+        );
```
Comment (sfdrogojan) [inline test/auth/AuthServiceUnitTests.spec.js]: Format

File: test/auth/AuthServiceUnitTests.spec.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        it('should execute the callApi method when the cached value for a cache key is expired', async () => {
+            let authService = new AuthService(clientConfiguration, apiClientStub, cacheService);
+
+            await authService.getTokenResponse();
+            let cachedDataExpirationCause = (tokenResponseData.data.expires_in - cacheService.invalidCacheWindowInSeconds) * 1000;
```
Comment (sfdrogojan) [inline test/auth/AuthServiceUnitTests.spec.js]: Can the .data prop be removed?

Comment (sfdrogojan) [inline test/auth/AuthServiceUnitTests.spec.js]: Could be renamed to validCacheWindowInMs for example

File: test/auth/CacheServiceUnitTests.spec.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
+const CacheService = require('../../src/Auth/CacheService');
+const TestHelper = require('./TestHelper');
+
+const expect = require('expect.js');
+const sinon = require('sinon');
+
+describe('CacheService', function () {
+    const cacheKey = 'cacheKey';
+    const date = 1568972613184;
```
Comment (sfdrogojan) [inline test/auth/CacheServiceUnitTests.spec.js]: For more readability, you can use for example new Date(2000, 1, 1).getTime()

File: test/auth/CacheServiceUnitTests.spec.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    afterEach(() => {
+        CacheService.cachedData = {};
+        clock.restore();
+    });
+
+    describe('get', function () {
+        it('should return null when passed parameter is not a key in the cache', function () {
+            expect(cacheService.get(cacheKey)).to.eql(null);
+            }
+        );
```
Comment (sfdrogojan) [inline test/auth/CacheServiceUnitTests.spec.js]: Format

File: test/auth/CacheServiceUnitTests.spec.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            expect(cachedTokenResponseData.tokenResponse.expires_in).to.be.greaterThan(0);
+            expect(cachedTokenResponseData.tokenResponse.rest_instance_url).to.eql('rest_instance_url');
+            expect(cachedTokenResponseData.tokenResponse.soap_instance_url).to.eql('soap_instance_url');
+            expect(cachedTokenResponseData.tokenResponse.scope).to.eql('scope');
+            expect(cachedTokenResponseData.tokenResponse.token_type).to.eql('Bearer');
+            }
+        );
+        it('should return null when the passed parameter is a cache key and the corresponding cache is expired', function () {
+            cacheService.addOrUpdate(cacheKey, tokenResponseData.data);
+            let cachedDataExpirationCause = (tokenResponseData.data.expires_in - cacheService.invalidCacheWindowInSeconds) * 1000;
```
Comment (sfdrogojan) [inline test/auth/CacheServiceUnitTests.spec.js]: Could be renamed to validCacheWindowInMs for example

File: test/auth/TestHelper.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+const ClientConfiguration = require('../../src/Auth/ClientConfiguration');
+
+class TestHelper {
+    static createTokenResponseData(){
+        return {
+            data: {
```
Comment (sfdrogojan) [inline test/auth/TestHelper.js:6]: data prop can be removed

**Review by sfdrogojan — APPROVED** [review]

**Review by sfdrogojan — APPROVED** [review]

File: src/Auth/CacheService.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
+class CacheService {
+    static cachedData = {};
+
+    constructor() {
+        this.invalidCacheWindowInSeconds = 30;
```
Comment (ghost) [inline src/Auth/CacheService.js]: you use this only in the get() method. no need to declare a variable in this case, just use the value in that method

**Review by ghost — CHANGES_REQUESTED** [review]
see my comments

**Review by ghost — APPROVED** [review]

**Pull request merged**